File: backend_js/src/users/domain/entities/user.entity.ts
import { BaseEntity } from "../../../core/entities/base.entity";

type basicUserInfo = {
    name: string;
    email: string;
    username: string;
};


export class UserEntity extends BaseEntity
{
    public name: string;
    public email: string;
    public password: string;
    public username: string;


    public updateDetails(data: basicUserInfo): void
    {
        this.name = data.name;
        this.email = data.email;
        this.username = data.username;
    }

    public updateName(newName: string): void
    {
        this.name = newName;
        this.updated_at = new Date();
    }

    public updateUsername(newUsername: string): void
    {
        this.username = newUsername;
        this.updated_at = new Date();
    }

    public updateEmail(newEmail: string): void
    {
        this.email = newEmail;
        this.updated_at = new Date();
    }

    public updatePassword(newPassword: string): void
    {
        this.password = newPassword;
        this.updated_at = new Date();
    }

    public markAsDeleted(): void
    {
        this.is_deleted = true;
        this.updated_at = new Date();
    }

    public restore(): void 
    {
        this.is_deleted = false;
        this.updated_at = new Date();
    }
}

File: backend_python/src/domain/entities/user_entity.py
# src/domain/entities/user.py
from datetime import datetime
from .base_entity import BaseEntity
from dataclasses import dataclass

@dataclass
class User(BaseEntity):
    name: str = ""
    email: str = ""
    password: str = ""
    username: str = ""
    sci_role: str = "" # e.g. 'metalurgist', 'geologist', etc.

    def update_details(self, name: str, email: str, username: str):
        self.name = name
        self.email = email
        self.username = username
        self.updated_at = datetime.now()

    def mark_as_deleted(self):
        self.is_deleted = True
        self.updated_at = datetime.now()